nixos/modules/hardware/printers.nix

```diff
@@ -0,0 +1,111 @@
+  ensurePrinter = p: ''
+    ${pkgs.cups}/bin/lpadmin -p '${p.name}' -E ${optionalString (p.location != null) "-L '${p.location}'"} ${optionalString (p.description != null) "-D '${p.description}'"} -v '${p.deviceUri}' -m '${p.model}'
+  '';
+  ensureDefaultPrinter = name: ''
+    ${pkgs.cups}/bin/lpoptions -d '${name}'
+  '';
+
+  forbiddenChars  = str: filter (c: c == "#" || c == "/") (stringToCharacters str) == [];
+  # TODO: is it possible to combine the forbiddenChars with that regex?
+  # test by using `nix repl` and then `builtins.match "[[:graph:]]+" "teststring#"`
+  printerName = (types.addCheck (types.strMatching "[[:graph:]]+") forbiddenChars)
+    // { description = "printable string without spaces, # and /"; };
+
+
+in
+{
+  options = {
+    hardware.printers = {
+      ensureDefaultPrinter = mkOption {
+        type = types.nullOr printerName;
+        default = null;
+        description = ''
+          Ensures the named printer is the default CUPS printer / printer queue.
+        '';
+      };
+      ensurePrinters = mkOption {
+        description = ''
+          Will regularly ensure that the given CUPS printers are configured as declared here.
+          If a printer's options are manually changed afterwards, they will be overwritten eventually.
+          This option will never delete any printer, even if removed from this list.
+          You can check existing printers with <literal>lpstat -s</literal>
+          and remove printers with <literal>lpadmin -x \<printer-name\></literal>.
+          Printers not listed here can still be manually configured.
+        '';
+        type = types.listOf (types.submodule {
+          options = {
+            name = mkOption {
+              type = printerName;
+              example = "BrotherHL_Workroom";
+              description = ''
+                Name of the printer / printer queue.
+                May contain any printable characters except "/", "#", and space.
+              '';
+            };
+            location = mkOption {
+              type = types.nullOr types.str;
+              default = null;
+              example = "Workroom";
+              description = ''
+                Optional human-readable location.
+              '';
+            };
+            description = mkOption {
+              type = types.nullOr types.str;
+              default = null;
+              example = "Brother HL-5140";
+              description = ''
+                Optional human-readable description.
+              '';
+            };
+            deviceUri = mkOption {
+              type = types.str;
+              example = [
+                "ipp://printserver.local/printers/BrotherHL_Workroom"
+                "usb://HP/DESKJET%20940C?serial=CN16E6C364BH"
+              ];
+              description = ''
+                How to reach the printer.
+                <literal>lpinfo -v</literal> shows a list of supported device URIs and schemes.
```

Comment:
I assume `lpinfo -v` can be used to figure out these URI's? In any case, in current master, `lpinfo -v` gives me an error:
```
$ lpinfo -v
lpinfo: Bad file descriptor
```

Reply:
Unsure about this, on 19.03 with a running cups, this just prints
```
lpinfo -v
network beh
network http
network ipps
network ipp
network lpd
network https
network socket
network smb
```
don't have master available to test though.

Reply:
couldn't you check like?
```
nix run -f channel:nixos-unstable  cups
lpinfo -v
```

Reply:
Did not know you could select another channel like that, thank you. I could confirm that combining the unstable `lpinfo` with stable cups service works.

The problem is the `Bad file descriptor` means that the `lpinfo` client can't communicate with the cups service, which probably means the service has some issues, then. It might just be though that @Infinisil 's cups service was just not running. (And in any way, this issue would be unrelated to this PR.)